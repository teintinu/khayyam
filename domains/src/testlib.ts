import { createJobManager, JobCallbackArgs, JobManager, Job, Node } from './job'
import { asap, defer } from 'pjobs'
import { Bundler } from './bundler'
import { ByPackage, createWorkspace, Package, WalkedJobs } from './workspace'
import { createProgress } from './progress'
import { System } from './sys'

export const cwd = process.cwd()
const fakeProcess = './scripts/fake-process.js'
const shell = true

export type TreeOfNode = [
  job: string,
  dependencies: TreeOfNode[],
]

export type Logger=ReturnType<typeof createFakeLog>

export function createFakeLog (verbose = false) {
  const logged: string[] = []
  const aOk = defer<void>()
  const bOk = defer<void>()
  const cOk = defer<void>()
  const handleOutput = jest.fn<void, JobCallbackArgs>()
    .mockImplementation(async (job, chunk, error) => {
      const lines = chunk.trim().split('\n')
      lines.forEach((s) => {
        if (s === 'a' || s === 'INPUT: a') setTimeout(aOk.resolve, 100)
        if (s === 'b' || s === 'INPUT: b') setTimeout(bOk.resolve, 100)
        if (s === 'c' || s === 'INPUT: c') setTimeout(cOk.resolve, 100)
        job.progress.update({
          state: 'working',
          message: (error ? ' (error): ' : '') + s
        })
      })
    })
  const jobManager = createJobManager()
  const fsListeners = new Set<()=>void>()
  let tmChanged: any
  const fakeSys: System & { simulateChange():void} = {
    workspaceChanged: fsChanged,
    listenForWorkspaceChanges (callback) {
      fsListeners.add(callback)
      return () => {
        fsListeners.delete(callback)
      }
    },
    getRepository: fsGetRepository,
    simulateChange: fsSimulateChange,
    loadWorkspace,
    notify: fsNotify
  }
  const fakeLogger = {
    jobManager,
    handleOutput,
    aOk,
    bOk,
    cOk,
    fakeSys,
    verbose,
    log (...args: string[]) {
      logged.push(args.join(' '))
      if (verbose) { console.info(args.join(' ')) }
    },
    logged,
    tree () {
      const done: {[id: number]:boolean} = {}
      const tree = jobManager.getTree()
      return tree.map(treeOfNode)
      function treeOfNode (node: Node): TreeOfNode {
        const id = node.job.id
        if (done[id]) {
          return [
            node.job.title + '(rec)',
            []
          ]
        }
        done[id] = true
        const n: TreeOfNode = [
          node.job.title + '=>[' +
          node.dependents.map(d => d.job.title).join() + ']',
          node.dependencies.map(treeOfNode)
        ]
        done[id] = false
        return n
      }
    }
  }
  return fakeLogger
  function fsChanged (): void {
    clearTimeout(tmChanged)
    tmChanged = setTimeout(() => {
      fsListeners.forEach((fn) => asap(fn))
    }, 1000)
  }
  function fsGetRepository (folder: string): string|undefined {
    return 'repo:' + folder
  }
  function fsSimulateChange ():void {
    fsChanged()
  }
  function loadWorkspace (folder:string) {
    const [sBundler, sPkg, sDeps, sLayer, sInvalid] = folder.split('/')
    let bundler:Bundler
    if (sBundler === 'npm') bundler = createFakeBundlerNPM(fakeLogger)
    else throw new Error('fake-fs-invalid-bundle')
    return createFakeWorkspace({
      logger: fakeLogger,
      bundler,
      packages: sPkg as FakePackages,
      deps: sDeps === 'deps',
      layers: sLayer === 'layers',
      invalid: sInvalid === 'invalid'
    })
  }
  function fsNotify (msg: string, pkgName: string) {
    fakeLogger.log('fsNotify:' + msg + ' on package: ' + pkgName)
  }
}

export function createFakeProgress (logger: Logger, title: string, manual: boolean) {
  const pg = createProgress(manual)
  pg.on((status) => {
    logger.log(title + ' ' + status.state + ' ' + status.message)
  })
  return pg
}

export function createFakeJob (logger: Logger, delay: number, title: string, args: string[], manual: boolean = false) {
  logger.log('creating job [' + title + ']')
  const job = logger.jobManager.createJob({
    title,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...args],
    shell,
    queue: 'sequential',
    manual
  })
  job.progress.on((status) => {
    logger.log('pg ' + title + ': ' + status.state + ' ' + status.message)
  })
  job.listen(logger.handleOutput)
  return job
}

export function create2FakeJob (logger: Logger, delay: number, mainTitle: string, mainArgs: string[], depTitle: string, depArgs: string[]) {
  logger.log('creating main job [' + mainTitle + ']')
  const main = logger.jobManager.createJob({
    title: mainTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...mainArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  logger.log('creating dep job [' + depTitle + ']')
  const dep = logger.jobManager.createJob({
    title: depTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...depArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  main.depends(dep)
  main.progress.on((status) => {
    logger.log('pg-main ' + status.state + ' ' + status.message)
  })
  dep.progress.on((status) => {
    logger.log('pg-dep ' + status.state + ' ' + status.message)
  })
  main.listen(logger.handleOutput)
  dep.listen(logger.handleOutput)
  return {
    main, dep
  }
}

function createFakeBundlerNPM (logger:Logger) {
  const fakeBundler: Bundler = {
    name: 'fakeBundlerNPM',
    watch (pkg) {
      return ['fakeBundlerNPM']
    },
    build (pkg, jobManager, goal) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-build-' + goal)
    },
    test (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-test')
    },
    serve (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-serve')
    },
    lint (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-lint')
    },
    publish (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-publish')
    }
  }
  return fakeBundler
  function createFakeBundlerCommand (pkg: Package, cmdJobManager: JobManager, cmd: string) {
    expect(cmdJobManager).toBe(logger.jobManager)
    return createFakeJob(
      logger,
      100,
      pkg.name + '(' + cmd + ')',
      [
        cmd
      ]
    )
  }
}

type FakePackages='x'|'xy'|'abcde'

function createFakeWorkspace (
  { logger, bundler, packages, deps, layers, invalid }:
  { logger: Logger, bundler: Bundler, packages:FakePackages, deps: boolean; layers: boolean; invalid: boolean }) {
  const pkgs:{
      [n:string]:Package
    } = {}
  for (const n of packages) {
    const pkg: Readonly<Package > = {
      name: n,
      layer: layers ? n : '',
      dependencies: [],
      bundlers: [bundler.name],
      folder: n
    }
    pkgs[n] = pkg
  }
  if (deps) {
    if (packages === 'xy') {
      pkgs.x = {
        ...pkgs.x,
        dependencies: invalid ? ['z'] : ['y']
      }
    }
    if (packages === 'abcde') {
      pkgs.a = {
        ...pkgs.a,
        dependencies: invalid ? ['b', 'c', 'z'] : ['b', 'c']
      }
      pkgs.b = {
        ...pkgs.b,
        dependencies: ['c', 'd']
      }
      pkgs.c = {
        ...pkgs.c,
        dependencies: ['d']
      }
      pkgs.d = {
        ...pkgs.d,
        dependencies: ['e']
      }
    }
  }
  const ws = createWorkspace({
    sys: logger.fakeSys,
    layers: [],
    packages: Object.keys(pkgs).map((n) => pkgs[n]),
    bundlers: [bundler]
  })
  return ws
}

export function jobDebug (job: Job) {
  return job.title + '(' +
    [...job.dependencies]
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(dep => dep.title) +
    ')'
}

export function walkedDebug (walkJobs: WalkedJobs) {
  const ret: ByPackage<unknown> = {}
  const pkgs = Object.keys(walkJobs.jobs).sort()
  pkgs.forEach((pkg) => {
    ret[pkg] = ([...walkJobs.jobs[pkg]] || [])
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(jobDebug)
  })
  return ret
}
