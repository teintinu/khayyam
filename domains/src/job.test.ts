import { create2FakeJob, createFakeJob, createFakeLog } from './testlib'

describe('job', () => {
  const delay = 150
  it('without dependencies', async () => {
    const logger = createFakeLog()
    const job = createFakeJob(logger, delay, 'job', ['a'])
    expect(job.title).toBe('job')
    expect(job.progress.actual.state).toBe('created')
    expect(logger.handleOutput).not.toBeCalled()
    expect(logger.logged?.length).toBe(0)
    const p = job.execute()
    await logger.aOk.promise
    expect(job.progress.actual.state).toBe('working')
    expect(logger.logged).toMatchSnapshot('01:log-working')
    expect(logger.tree()).toMatchSnapshot('02:tree-working')
    expect(job.progress.actual.state).toBe('working')
    expect(job.progress.actual.message).toBe('a')
    expect(await p).toBeUndefined()
    expect(job.progress.actual.state).toBe('finished')
    expect(logger.logged).toMatchSnapshot('03:log-finished')
    expect(logger.tree()).toMatchSnapshot('04:tree-finished')
  })
  it('multiline', async () => {
    const logger = createFakeLog()
    const job = createFakeJob(logger, delay, 'multi', ['a', 'b'])
    expect(job.title).toBe('multi')
    expect(job.progress.actual.state).toBe('created')
    expect(logger.handleOutput).not.toBeCalled()
    expect(logger.logged?.length).toBe(0)
    const p = job.execute()
    await logger.aOk.promise
    expect(logger.logged).toMatchSnapshot('01:log-aOk')
    expect(logger.tree()).toMatchSnapshot('02:tree-aOk')
    expect(job.progress.actual.state).toBe('working')
    expect(job.progress.actual.message).toBe('a')
    await logger.bOk.promise
    expect(logger.logged).toMatchSnapshot('03:log-bOk')
    expect(logger.tree()).toMatchSnapshot('04:tree-bOk')
    expect(job.progress.actual.state).toBe('working')
    expect(job.progress.actual.message).toBe('b')
    expect(await p).toBeUndefined()
    expect(job.progress.actual.state).toBe('finished')
    expect(logger.logged).toMatchSnapshot('05:log-finished')
    expect(logger.tree()).toMatchSnapshot('06:tree-finished')
  })
  it('typing', async () => {
    const logger = createFakeLog()
    const job = createFakeJob(
      logger,
      delay,
      'typing',
      [
        'READ',
        'a',
        'READ',
        'b',
        'READ',
        'c'
      ]
    )
    expect(job.title).toBe('typing')
    expect(job.progress.actual.state).toBe('created')
    expect(logger.handleOutput).not.toBeCalled()
    expect(logger.logged?.length).toBe(0)
    const p = job.execute()
    await logger.aOk.promise
    expect(logger.logged).toMatchSnapshot('01:log-aOk')
    expect(logger.tree()).toMatchSnapshot('02:tree-aOk')
    job.type('123\n')
    await logger.bOk.promise
    expect(logger.logged).toMatchSnapshot('03:log-bOk')
    expect(logger.tree()).toMatchSnapshot('04:tree-bOk')
    job.type('456\n')
    await logger.cOk.promise
    expect(logger.logged).toMatchSnapshot('05:log-cOk')
    expect(logger.tree()).toMatchSnapshot('06:tree-cOk')
    job.type('789\n')
    expect(await p).toBeUndefined()
    expect(logger.logged).toMatchSnapshot('07:log-finished')
    expect(logger.tree()).toMatchSnapshot('08:tree-finished')
    expect(job.progress.actual.state).toBe('finished')
  })
  it('with dependencies', async () => {
    const logger = createFakeLog()
    const { main, dep } = create2FakeJob(
      logger,
      delay,
      'main', ['READ', 'a'],
      'dep', ['READ', 'b']
    )
    expect(main.title).toBe('main')
    expect(dep.title).toBe('dep')
    expect(main.progress.actual.state).toBe('created')
    expect(dep.progress.actual.state).toBe('created')
    expect(logger.handleOutput).not.toBeCalled()
    expect(logger.logged?.length).toBe(0)
    const pMain = main.execute()
    const pDep = dep.execute()
    await logger.bOk.promise
    expect(dep.progress.actual.state).toBe('working')
    expect(main.progress.actual.state).toBe('waiting')
    expect(logger.logged).toMatchSnapshot('01:log-dep working')
    expect(logger.tree()).toMatchSnapshot('02:tree-dep working')
    dep.type('dep-ok\n')
    await logger.aOk.promise
    expect(dep.progress.actual.state).toBe('finished')
    expect(main.progress.actual.state).toBe('working')
    expect(logger.logged).toMatchSnapshot('03:log-dep finished')
    expect(logger.tree()).toMatchSnapshot('04:tree-dep finished')
    main.type('main-ok\n')
    expect(await pDep).toBeUndefined()
    expect(await pMain).toBeUndefined()
    expect(dep.progress.actual.state).toBe('finished')
    expect(main.progress.actual.state).toBe('finished')
    expect(logger.logged).toMatchSnapshot('05:log-dep working')
    expect(logger.tree()).toMatchSnapshot('06:tree-dep working')
  })

  it('auto-start dependencies', async () => {
    const logger = createFakeLog()
    const { main, dep } = create2FakeJob(
      logger,
      delay,
      'main', ['READ', 'a'],
      'dep', ['READ', 'b']
    )
    expect(main.title).toBe('main')
    expect(dep.title).toBe('dep')
    expect(main.progress.actual.state).toBe('created')
    expect(dep.progress.actual.state).toBe('created')
    expect(logger.handleOutput).not.toBeCalled()
    expect(logger.logged?.length).toBe(0)
    const p1 = main.execute()
    const p2 = dep.execute()
    await logger.bOk.promise
    expect(dep.progress.actual.state).toBe('working')
    expect(main.progress.actual.state).toBe('waiting')
    expect(logger.logged).toMatchSnapshot('01:log-dep working')
    expect(logger.tree()).toMatchSnapshot('02:tree-dep working')
    dep.type('dep-ok\n')
    await logger.aOk.promise
    expect(dep.progress.actual.state).toBe('finished')
    expect(main.progress.actual.state).toBe('working')
    expect(logger.logged).toMatchSnapshot('03:log-dep finished')
    expect(logger.tree()).toMatchSnapshot('04:tree-dep finished')
    main.type('main-ok\n')
    expect(await p1).toBeUndefined()
    expect(await p2).toBeUndefined()
    expect(dep.progress.actual.state).toBe('finished')
    expect(main.progress.actual.state).toBe('finished')
    expect(logger.logged).toMatchSnapshot('05:log-finished')
    expect(logger.tree()).toMatchSnapshot('06:tree-finished')
  })
})
