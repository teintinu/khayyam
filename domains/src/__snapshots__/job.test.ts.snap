// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`job auto-start dependencies: 00:create 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
]
`;

exports[`job auto-start dependencies: 01:log-dep working 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
]
`;

exports[`job auto-start dependencies: 02:tree-dep working 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job auto-start dependencies: 03:log-dep finished 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
  "pg-dep working b: dep-ok",
  "pg-dep working finished",
  "pg-dep finished undefined",
  "pg-main working undefined",
  "pg-main working started",
  "pg-main working INPUT: a",
]
`;

exports[`job auto-start dependencies: 04:tree-dep finished 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job auto-start dependencies: 05:log-finished 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
  "pg-dep working b: dep-ok",
  "pg-dep working finished",
  "pg-dep finished undefined",
  "pg-main working undefined",
  "pg-main working started",
  "pg-main working INPUT: a",
  "pg-main working a: main-ok",
  "pg-main working finished",
  "pg-main finished undefined",
]
`;

exports[`job auto-start dependencies: 06:tree-finished 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job multiline: 00:create 1`] = `
Array [
  "creating job [multi]",
]
`;

exports[`job multiline: 01:log-aOk 1`] = `
Array [
  "creating job [multi]",
  "pg multi: waiting undefined",
  "pg multi: working undefined",
  "pg multi: working started",
  "pg multi: working a",
]
`;

exports[`job multiline: 02:tree-aOk 1`] = `
"multi
  dependents:
  dependencies:"
`;

exports[`job multiline: 03:log-bOk 1`] = `
Array [
  "creating job [multi]",
  "pg multi: waiting undefined",
  "pg multi: working undefined",
  "pg multi: working started",
  "pg multi: working a",
  "pg multi: working b",
]
`;

exports[`job multiline: 04:tree-bOk 1`] = `
"multi
  dependents:
  dependencies:"
`;

exports[`job multiline: 05:log-finished 1`] = `
Array [
  "creating job [multi]",
  "pg multi: waiting undefined",
  "pg multi: working undefined",
  "pg multi: working started",
  "pg multi: working a",
  "pg multi: working b",
  "pg multi: working finished",
  "pg multi: finished undefined",
]
`;

exports[`job multiline: 06:tree-finished 1`] = `
"multi
  dependents:
  dependencies:"
`;

exports[`job typing: 00:create 1`] = `
Array [
  "creating job [typing]",
]
`;

exports[`job typing: 01:log-aOk 1`] = `
Array [
  "creating job [typing]",
  "pg typing: waiting undefined",
  "pg typing: working undefined",
  "pg typing: working started",
  "pg typing: working INPUT: a",
]
`;

exports[`job typing: 02:tree-aOk 1`] = `
"typing
  dependents:
  dependencies:"
`;

exports[`job typing: 03:log-bOk 1`] = `
Array [
  "creating job [typing]",
  "pg typing: waiting undefined",
  "pg typing: working undefined",
  "pg typing: working started",
  "pg typing: working INPUT: a",
  "pg typing: working a: 123",
  "pg typing: working INPUT: b",
]
`;

exports[`job typing: 04:tree-bOk 1`] = `
"typing
  dependents:
  dependencies:"
`;

exports[`job typing: 05:log-cOk 1`] = `
Array [
  "creating job [typing]",
  "pg typing: waiting undefined",
  "pg typing: working undefined",
  "pg typing: working started",
  "pg typing: working INPUT: a",
  "pg typing: working a: 123",
  "pg typing: working INPUT: b",
  "pg typing: working b: 456",
  "pg typing: working INPUT: c",
]
`;

exports[`job typing: 06:tree-cOk 1`] = `
"typing
  dependents:
  dependencies:"
`;

exports[`job typing: 07:log-finished 1`] = `
Array [
  "creating job [typing]",
  "pg typing: waiting undefined",
  "pg typing: working undefined",
  "pg typing: working started",
  "pg typing: working INPUT: a",
  "pg typing: working a: 123",
  "pg typing: working INPUT: b",
  "pg typing: working b: 456",
  "pg typing: working INPUT: c",
  "pg typing: working c: 789",
  "pg typing: working finished",
  "pg typing: finished undefined",
]
`;

exports[`job typing: 08:tree-finished 1`] = `
"typing
  dependents:
  dependencies:"
`;

exports[`job with dependencies: 00:create 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
]
`;

exports[`job with dependencies: 01:log-dep working 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
]
`;

exports[`job with dependencies: 02:tree-dep working 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job with dependencies: 03:log-dep finished 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
  "pg-dep working b: dep-ok",
  "pg-dep working finished",
  "pg-dep finished undefined",
  "pg-main working undefined",
  "pg-main working started",
  "pg-main working INPUT: a",
]
`;

exports[`job with dependencies: 04:tree-dep finished 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job with dependencies: 05:log-dep working 1`] = `
Array [
  "creating main job [main]",
  "creating dep job [dep]",
  "pg-main waiting undefined",
  "pg-dep waiting undefined",
  "pg-dep working undefined",
  "pg-dep working started",
  "pg-dep working INPUT: b",
  "pg-dep working b: dep-ok",
  "pg-dep working finished",
  "pg-dep finished undefined",
  "pg-main working undefined",
  "pg-main working started",
  "pg-main working INPUT: a",
  "pg-main working a: main-ok",
  "pg-main working finished",
  "pg-main finished undefined",
]
`;

exports[`job with dependencies: 06:tree-dep working 1`] = `
"main
  dependents:
  dependencies:
    dep
      dependents:
        main
      dependencies:"
`;

exports[`job without dependencies: 00:create 1`] = `
Array [
  "creating job [job]",
]
`;

exports[`job without dependencies: 01:log-working 1`] = `
Array [
  "creating job [job]",
  "pg job: waiting undefined",
  "pg job: working undefined",
  "pg job: working started",
  "pg job: working a",
]
`;

exports[`job without dependencies: 02:tree-working 1`] = `
"job
  dependents:
  dependencies:"
`;

exports[`job without dependencies: 03:log-finished 1`] = `
Array [
  "creating job [job]",
  "pg job: waiting undefined",
  "pg job: working undefined",
  "pg job: working started",
  "pg job: working a",
  "pg job: working finished",
  "pg job: finished undefined",
]
`;

exports[`job without dependencies: 04:tree-finished 1`] = `
"job
  dependents:
  dependencies:"
`;
